Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -221,7 +209,18 @@ public int add(double[] point, long sequenceNum) {
             // point has to be written, otherwise we need to write the full shingle
             int amountToWrite = checkShingleAlignment(startOfFreeSegment, tempPoint) ? baseDimension : dimensions;
 
-            verifyAndMakeSpace(amountToWrite);
+            if (startOfFreeSegment > currentStoreCapacity * dimensions - amountToWrite) {
+                // try compaction and then resizing
+                compact();
+                // the compaction can change the array contents
+                amountToWrite = checkShingleAlignment(startOfFreeSegment, tempPoint) ? baseDimension : dimensions;
+                if (startOfFreeSegment > currentStoreCapacity * dimensions - amountToWrite) {
+                    checkState(dynamicResizingEnabled, " out of store, enable dynamic resizing ");
+                    resizeStore();
+                    checkState(startOfFreeSegment + amountToWrite <= currentStoreCapacity * dimensions, "out of space");
```

Comment:
When would this happen? Is there a max size that the point store can grow to?

Reply:
Yes. The maximum size is "capacity * dimensions" (and twice that if shingle rotations are enabled) ; if dynamicResizing is turned on then currentStoreCapacity starts small and can reach capacity. However that depends on the timeDecay factor and it is possible that the pointStore never reaches full capacity -- for example if the time decay is large, we are essentially ensuring a "sliding window" (last few data points) behavior and that would determine the pointStore (and maybe a few rare straggler points).

Reply:
Could you explain why shingle rotation need twice the max capacity? I thought one is a cyclic buffer and another is a sliding window. The buffer and window's size are the same, right?

Also, if shingle rotation is enabled, should you change the check condition to startOfFreeSegment + amountToWrite <= currentStoreCapacity * dimensions * 2 ?